File: plugins/plugin-scratchpad/python/elizaos_plugin_scratchpad/actions/append.py
"""SCRATCHPAD_APPEND action - Append content to an existing scratchpad entry."""

from __future__ import annotations

import logging
import re
from dataclasses import dataclass
from typing import TYPE_CHECKING

from elizaos_plugin_scratchpad.service import create_scratchpad_service
from elizaos_plugin_scratchpad.types import ScratchpadWriteOptions

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State

logger = logging.getLogger(__name__)

EXTRACT_TEMPLATE = """Extract the scratchpad entry ID and content to append from the user's message.

User message: {{text}}

Available scratchpad entries:
{{entries}}

Respond with XML containing:
- id: The ID of the scratchpad entry to append to (required)
- content: The new content to append (required)

<response>
<id>entry-id</id>
<content>Content to append</content>
</response>"""


def _parse_xml_field(text: str, field_name: str) -> str | None:
    """Extract a field value from XML-like response text."""
    match = re.search(rf"<{field_name}>([\s\S]*?)</{field_name}>", text)
    return match.group(1).strip() if match else None


@dataclass
class AppendResult:
    """Result of the SCRATCHPAD_APPEND action."""

    success: bool
    text: str
    error: str | None = None


async def _extract_append_info(
    runtime: IAgentRuntime,
    message: Memory,
    available_entries: str,
) -> dict[str, str] | None:
    """Use the LLM to extract the entry ID and content to append."""
    message_text = message.content.text if message.content else ""
    prompt = EXTRACT_TEMPLATE.replace("{{text}}", message_text).replace(
        "{{entries}}", available_entries
    )

    result = await runtime.use_model("TEXT_SMALL", {"prompt": prompt, "stopSequences": []})
    result_str = str(result)
    logger.debug("[ScratchpadAppend] Extract result: %s", result_str)

    entry_id = _parse_xml_field(result_str, "id")
    content = _parse_xml_field(result_str, "content")

    if not entry_id or not content:
        logger.error("[ScratchpadAppend] Failed to extract valid append info")
        return None

    return {"id": entry_id, "content": content}


async def handle_scratchpad_append(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
    options: HandlerOptions | None = None,
    callback: HandlerCallback | None = None,
    responses: list[Memory] | None = None,
) -> AppendResult | None:
    """Handle the SCRATCHPAD_APPEND action."""
    service = create_scratchpad_service(runtime)

    entries = await service.list()
    entries_context = "\n".join(f'- {e.id}: "{e.title}"' for e in entries)

    if not entries:
        msg = "There are no scratchpad entries to append to. Create one first with SCRATCHPAD_WRITE."
        if callback:
            await callback(
                {
                    "text": msg,
                    "actions": ["SCRATCHPAD_APPEND_EMPTY"],
                    "source": message.content.source if message.content else None,
                }
            )
        return AppendResult(success=False, text="No entries available")

    append_info = await _extract_append_info(runtime, message, entries_context)

    if not append_info:
        msg = f"I couldn't determine which note to update or what to add. Available entries:\n{entries_context}"
        if callback:
            await callback(
                {
                    "text": msg,
                    "actions": ["SCRATCHPAD_APPEND_FAILED"],
                    "source": message.content.source if message.content else None,
                }
            )
        return AppendResult(success=False, text="Failed to extract append info")

    try:
        entry_id = append_info["id"]
        new_content = append_info["content"]

        exists = await service.exists(entry_id)
        if not exists:
            msg = f'Scratchpad entry "{entry_id}" not found. Available entries:\n{entries_context}'
            if callback:
                await callback(
                    {
                        "text": msg,
                        "actions": ["SCRATCHPAD_APPEND_NOT_FOUND"],
                        "source": message.content.source if message.content else None,
                    }
                )
            return AppendResult(success=False, text="Entry not found")

        existing_entry = await service.read(entry_id)

        entry = await service.write(
            existing_entry.title,
            new_content,
            ScratchpadWriteOptions(append=True, tags=existing_entry.tags or None),
        )

        success_msg = f'Successfully appended content to "{entry.title}" ({entry.id}).'
        if callback:
            await callback(
                {
                    "text": success_msg,
                    "actions": ["SCRATCHPAD_APPEND_SUCCESS"],
                    "source": message.content.source if message.content else None,
                }
            )

        return AppendResult(success=True, text=success_msg)

    except Exception as exc:
        error_msg = f"Failed to append to the note: {exc}"
        logger.error("[ScratchpadAppend] Error: %s", exc)
        if callback:
            await callback(
                {
                    "text": error_msg,
                    "actions": ["SCRATCHPAD_APPEND_FAILED"],
                    "source": message.content.source if message.content else None,
                }
            )
        return AppendResult(
            success=False, text="Failed to append to scratchpad entry", error=str(exc)
        )


async def validate_scratchpad_append(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> bool:
    """Always validate - selection is handled in the handler."""
    return True


SCRATCHPAD_APPEND_ACTION: dict[str, object] = {
    "name": "SCRATCHPAD_APPEND",
    "similes": [
        "ADD_TO_NOTE",
        "APPEND_NOTE",
        "UPDATE_NOTE",
        "EXTEND_NOTE",
    ],
    "description": "Append additional content to an existing scratchpad entry.",
    "examples": [
        [
            {
                "name": "{{name1}}",
                "content": {
                    "text": "Add to my meeting-notes: Action item - follow up with design team"
                },
            },
            {
                "name": "{{name2}}",
                "content": {
                    "text": 'Successfully appended content to "Meeting Notes" (meeting-notes).',
                    "actions": ["SCRATCHPAD_APPEND_SUCCESS"],
                },
            },
        ],
    ],
}

File: packages/core/src/features/advanced-capabilities/clipboard/actions/append.ts
import {
	type Action,
	type HandlerCallback,
	type HandlerOptions,
	type IAgentRuntime,
	logger,
	type Memory,
	type State,
} from "../../../../types/index.ts";
import { hasActionContextOrKeyword } from "../../../../utils/action-validation.ts";
import { createClipboardService } from "../services/clipboardService.ts";
import { requireActionSpec } from "../specs.ts";

interface AppendInput {
	id: string;
	content: string;
}

function isValidAppendInput(obj: Record<string, unknown>): boolean {
	return (
		typeof obj.id === "string" &&
		obj.id.length > 0 &&
		typeof obj.content === "string" &&
		obj.content.length > 0
	);
}

function readParams(options?: HandlerOptions): Record<string, unknown> {
	return options?.parameters && typeof options.parameters === "object"
		? (options.parameters as Record<string, unknown>)
		: {};
}

function extractAppendInfo(
	message: Memory,
	options?: HandlerOptions,
): AppendInput | null {
	const params = readParams(options);
	const raw = {
		id: params.id ?? message.content.id ?? message.content.entryId,
		content: params.content ?? message.content.content,
	};

	if (!isValidAppendInput(raw)) {
		logger.error("[ClipboardAppend] Failed to extract valid append info");
		return null;
	}

	return {
		id: String(raw.id),
		content: String(raw.content),
	};
}

const spec = requireActionSpec("CLIPBOARD_APPEND");
const MAX_CONTEXT_ENTRIES = 20;
const MAX_TITLE_CHARS = 120;
const MAX_APPEND_CHARS = 12000;

function truncateText(text: string, max: number): string {
	return text.length <= max ? text : `${text.slice(0, max)}\n…[truncated]`;
}

export const clipboardAppendAction: Action = {
	name: spec.name,
	contexts: ["files", "knowledge", "agent_internal"],
	roleGate: { minRole: "ADMIN" },
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,

	validate: async (
		_runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		options?: HandlerOptions,
	): Promise<boolean> => {
		const __avParams = readParams(options);
		if (isValidAppendInput(__avParams)) {
			return true;
		}
		return hasActionContextOrKeyword(message, state, {
			contexts: ["files", "knowledge", "agent_internal"],
			keywords: ["clipboard", "append", "add to note", "add to clipboard"],
		});
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		_stateFromTrigger: State | undefined,
		_options: HandlerOptions | undefined,
		callback?: HandlerCallback,
		_responses?: Memory[],
	) => {
		const service = createClipboardService(runtime);

		// Get list of available entries for context
		const entries = await service.list();
		const entriesContext = entries
			.slice(0, MAX_CONTEXT_ENTRIES)
			.map((e) => `- ${e.id}: "${e.title}"`)
			.join("\n");
		const omittedContext =
			entries.length > MAX_CONTEXT_ENTRIES
				? `\n…${entries.length - MAX_CONTEXT_ENTRIES} more entries omitted.`
				: "";

		if (entries.length === 0) {
			if (callback) {
				await callback({
					text: "There are no clipboard entries to append to. Create one first with CLIPBOARD_WRITE.",
					actions: ["CLIPBOARD_APPEND_EMPTY"],
					source: message.content.source,
				});
			}
			return { success: false, text: "No entries available" };
		}

		const appendInfo = extractAppendInfo(message, _options);

		if (!appendInfo) {
			if (callback) {
				await callback({
					text: `I couldn't determine which note to update or what to add. Available entries:\n${entriesContext}${omittedContext}`,
					actions: ["CLIPBOARD_APPEND_FAILED"],
					source: message.content.source,
				});
			}
			return { success: false, text: "Failed to extract append info" };
		}

		try {
			// Check if entry exists
			const exists = await service.exists(appendInfo.id);
			if (!exists) {
				if (callback) {
					await callback({
						text: `Clipboard entry "${appendInfo.id}" not found. Available entries:\n${entriesContext}${omittedContext}`,
						actions: ["CLIPBOARD_APPEND_NOT_FOUND"],
						source: message.content.source,
					});
				}
				return { success: false, text: "Entry not found" };
			}

			// Get existing entry to preserve title
			const existingEntry = await service.read(appendInfo.id);

			// Write with append option
			const entry = await service.write(
				existingEntry.title,
				truncateText(appendInfo.content, MAX_APPEND_CHARS),
				{
					append: true,
					tags: existingEntry.tags,
				},
			);

			const successMessage = `Successfully appended content to "${entry.title}" (${entry.id}).`;

			if (callback) {
				await callback({
					text: successMessage,
					actions: ["CLIPBOARD_APPEND_SUCCESS"],
					source: message.content.source,
				});
			}

			return {
				success: true,
				text: successMessage,
				data: {
					entryId: entry.id,
					title: truncateText(entry.title, MAX_TITLE_CHARS),
					appendedChars: Math.min(appendInfo.content.length, MAX_APPEND_CHARS),
					truncated: appendInfo.content.length > MAX_APPEND_CHARS,
				},
			};
		} catch (error) {
			const errorMsg = error instanceof Error ? error.message : String(error);
			logger.error("[ClipboardAppend] Error:", errorMsg);
			if (callback) {
				await callback({
					text: `Failed to append to the note: ${errorMsg}`,
					actions: ["CLIPBOARD_APPEND_FAILED"],
					source: message.content.source,
				});
			}
			return { success: false, text: "Failed to append to clipboard entry" };
		}
	},

	parameters: [
		{
			name: "id",
			description: "Clipboard entry ID to append content to.",
			required: true,
			schema: { type: "string" as const, minLength: 1 },
		},
		{
			name: "content",
			description: "Content to append to the clipboard entry.",
			required: true,
			schema: { type: "string" as const, minLength: 1 },
		},
	],
	examples: [],
};

export default clipboardAppendAction;
